### PR TITLE
Conditionally required case search fields: re-validate on field change

### PR DESCRIPTION
**Pull request by orangejenny**
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-2203

## Feature Flag
Case search

### Safety story
Some risk, limited to case search projects, will go through QA.

### Automated test coverage

None

### QA Plan
https://dimagi-dev.atlassian.net/browse/QA-4342

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 1]
```diff
@@ -434,38 +424,61 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
...
+                FormplayerFrontend.trigger("menu:query", self.getAnswers());
+            });
+        },
 
-            FormplayerFrontend.trigger('clearNotifications', errorHTML, true);
+        /*
+         *  Send request to formplayer to validate fields. Displays any errors.
+         *  Returns a promise that contains the formplayer response.
+         */
+        validateFields: function () {
```
Comment (esoergel) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:438]: Are you worried about this being called much more often than it needs to be?  I wonder if it'd be worth trying to figure out some logic around caching this and expiring when user input changes.  Sounds tricky.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 2]
```diff
@@ -434,38 +424,61 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             var self = this;
             e.preventDefault();
 
-            var invalidFields = [];
-            var errorHTML = gettext("Please check the following fields:");
+            // validateFields will likely already have been called when user blurred the last field,
+            // but call it here just in case they didn't fill anything out
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:428]: the javascript event (when a field loses focus)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:438]: I'm a little bit worried about it. I'm planning to put that worry out of my head until we get complaints (acknowledging that might not take long).

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 3]
```diff
@@ -204,6 +204,10 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             'click @ui.searchForBlank': 'toggleBlankSearch',
         },
 
+        _render: function () {
+            _.delay(this.render);
```
Comment (esoergel) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:208]: What does this accomplish - move it off thread?  Doesn't `_.delay` take a [`wait` arg](https://underscorejs.org/#delay)?

Comment (orangejenny) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js]: Yeah, the intent was not to render until the current call stack is complete (including closing the select2). Looking at the underscore docs, I meant to use `defer`, which should be equivalent but more readable. Updated in https://github.com/dimagi/commcare-hq/pull/31958/commits/3bf2625526eb7e56efe5a9085a53d5285590d384

**Review by esoergel — APPROVED** [review]

**Pull request merged**